Binary files grep-2.6.3.orig/src/.main.c.swp and grep-2.6.3/src/.main.c.swp differ
diff -uNrp grep-2.6.3.orig/src/main.c grep-2.6.3/src/main.c
--- grep-2.6.3.orig/src/main.c	2010-08-26 10:42:34 +0000
+++ grep-2.6.3/src/main.c	2010-08-26 13:07:38 +0000
@@ -860,10 +860,10 @@ prpending (char const *lim)
       size_t match_size;
       --pending;
       if (outleft
-	  || ((execute(lastout, nl + 1 - lastout,
+	  || ((execute(lastout, (nl?:lim) + 1 - lastout,
 		       &match_size, NULL) == (size_t) -1)
 	      == !out_invert))
-	prline (lastout, nl + 1, SEP_CHAR_REJECTED);
+	prline (lastout, (nl?nl + 1:lim), SEP_CHAR_REJECTED);
       else
 	pending = 0;
     }
@@ -908,7 +908,10 @@ prtext (char const *beg, char const *lim
       while (p < beg)
 	{
 	  char const *nl = memchr (p, eol, beg - p);
-	  nl++;
+	  if (nl)
+	    nl++;
+	  else
+	    nl = beg;
 	  prline (p, nl, SEP_CHAR_REJECTED);
 	  p = nl;
 	}
@@ -920,7 +923,10 @@ prtext (char const *beg, char const *lim
       for (n = 0; p < lim && n < outleft; n++)
 	{
 	  char const *nl = memchr (p, eol, lim - p);
-	  nl++;
+	  if (nl)
+	    nl++;
+	  else
+	    nl = lim;
 	  if (!out_quiet)
 	    prline (p, nl, SEP_CHAR_SELECTED);
 	  p = nl;
